### PR TITLE
Refactor ActionManagementDAO to handle action type specific properties 

### PR DESCRIPTION
**Pull request by ashanthamara**
### Proposed changes in this pull request

1. With this PR, ActionManagementDAO will be able to handle properties specific to each actionTypes. (Currently for `PRE_ISSUE_ACCESS_TOKEN` and `PRE_UPDATE_PASSWORD`)
2. action-mgt component will utilize certificate-mgt for the operations related to the certificates used in `PRE_UPDATE_PASSWORD` action type.
3. Also with this PR, the `updateActionEndpointAuthentication` will be removed as it has no use case in the product
4. The `IDN_ACTION_ENDPOINT` table will be renamed to `IDN_ACTION_PROPERTIES` and now it will be utilize to store action type related properties as well.

### Related Issue
- https://github.com/wso2/product-is/issues/21661

### CODE REVIEW

File: components/action-mgt/org.wso2.carbon.identity.action.management/pom.xml [hunk 1]
```diff
@@ -178,14 +183,14 @@
...
-                                            <minimum>0.70</minimum>
+                                            <minimum>0.77</minimum>
                                         </limit>
                                         <limit implementation="org.jacoco.report.check.Limit">
                                             <counter>COMPLEXITY</counter>
                                             <value>COVEREDRATIO</value>
                                                 <!-- Temp decreasing this. There is an ongoing effort to improve
                                                 the code complexity coverage.  -->
-                                            <minimum>0.60</minimum>
+                                            <minimum>0.68</minimum>
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/pom.xml:193]: Let's keep this to 0.7

File: components/action-mgt/org.wso2.carbon.identity.action.management/pom.xml [hunk 2]
```diff
@@ -178,14 +183,14 @@
                                             <!-- Temporarily decreasing this.
                                             Will be improved upon in a follow-up effort.
                                             Related issue: https://github.com/wso2/product-is/issues/21368 -->
-                                            <minimum>0.70</minimum>
+                                            <minimum>0.77</minimum>
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/pom.xml:186]: Target here should be 0.8

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/constant/ActionMgtConstants.java [hunk 3]
```diff
@@ -23,8 +23,10 @@
  */
 public class ActionMgtConstants {
 
-    public static final String URI_ATTRIBUTE = "uri";
-    public static final String AUTHN_TYPE_ATTRIBUTE = "authnType";
+    public static final String URI_PROPERTY = "uri";
+    public static final String AUTHN_TYPE_PROPERTY = "authnType";
+    public static final String PASSWORD_SHARING_FORMAT_PROPERTY = "passwordSharingFormat";
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/constant/ActionMgtConstants.java:28]: This constant should not come to the actions core. It's specific to a particular extension

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/constant/ActionMgtConstants.java [hunk 4]
```diff
@@ -23,8 +23,10 @@
...
+    public static final String CERTIFICATE_ID_PROPERTY = "certificateId";
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/constant/ActionMgtConstants.java:29]: Same here

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/constant/ActionMgtConstants.java [hunk 5]
```diff
@@ -53,19 +55,30 @@ public enum ErrorMessages {
                 "%s is empty."),
         ERROR_INVALID_ACTION_REQUEST_FIELD("60005", "Invalid request.",
                 "%s is invalid."),
+        ERROR_INVALID_ACTION_CERTIFICATE("60006", "Invalid request.", "Provided certificate is invalid."),
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/constant/ActionMgtConstants.java:58]: This is specific to a particular extension, we should be able to take these validations out from the core

File: components/action-mgt/org.wso2.carbon.identity.action.management/pom.xml (hunk 1 above)
Comment (malithie) [reply components/action-mgt/org.wso2.carbon.identity.action.management/pom.xml:193]: Let's keep the threshold to 0.6

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionBuilder.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+package org.wso2.carbon.identity.action.management;
+
+import org.wso2.carbon.identity.action.management.dao.model.ActionDTO;
+import org.wso2.carbon.identity.action.management.model.Action;
+
+/**
+ * This interface defines the Action Resolver.
+ * Action Resolver is the component that is responsible for the conversions between Action and ExtendedAction
+ * objects.
+ */
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionBuilder.java:28]: Can we say an Action Model converter

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionBuilder.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+public interface ActionBuilder {
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionBuilder.java:29]: Shall we rename as ActionConverter

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionBuilder.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    default ActionDTO buildActionDTO(Action action) {
+
+        return new ActionDTO.Builder()
+                .id(action.getId())
+                .type(action.getType())
+                .name(action.getName())
+                .description(action.getDescription())
+                .status(action.getStatus())
+                .endpoint(action.getEndpoint())
+                .properties(null)
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionBuilder.java:48]: Should we set this as null ? We can just skip setting this right ?

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/factory/ActionBuilderFactory.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * This class defines the Action Object Builder Factory.
+ * Action Object Builder Factory is the component that is responsible for providing the {@link ActionBuilder}
+ * based on the action type.
+ */
+public class ActionBuilderFactory {
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/factory/ActionBuilderFactory.java:32]: Let's rename this as wel as per the comment above

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/model/PreUpdatePasswordAction.java [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+ */
+
+package org.wso2.carbon.identity.action.management.model;
+
+import org.wso2.carbon.identity.certificate.management.model.Certificate;
+
+/**
+ * PreUpdatePasswordAction.
+ */
+public class PreUpdatePasswordAction extends Action {
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/model/PreUpdatePasswordAction.java:26]: We should move this model to the respective extension point

Comment (ashanthamara) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/model/PreUpdatePasswordAction.java:26]: Fixed

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionBuilder.java (hunk 7 above)
Comment (ashanthamara) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionBuilder.java]: Fixed with https://github.com/wso2/carbon-identity-framework/pull/6119/commits/73129ff409dc220674594045370c909805220d13

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionBuilder.java (hunk 6 above)
Comment (ashanthamara) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/ActionBuilder.java]: Fixed with https://github.com/wso2/carbon-identity-framework/pull/6119/commits/73129ff409dc220674594045370c909805220d13

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/constant/ActionMgtConstants.java (hunk 5 above)
Comment (ashanthamara) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/constant/ActionMgtConstants.java]: Fixed

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/constant/ActionMgtConstants.java (hunk 4 above)
Comment (ashanthamara) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/constant/ActionMgtConstants.java]: Fixed

File: components/action-mgt/org.wso2.carbon.identity.action.management/pom.xml (hunk 1 above)
Comment (ashanthamara) [reply components/action-mgt/org.wso2.carbon.identity.action.management/pom.xml]: Fixed

File: components/action-mgt/org.wso2.carbon.identity.action.management/pom.xml (hunk 2 above)
Comment (ashanthamara) [reply components/action-mgt/org.wso2.carbon.identity.action.management/pom.xml]: Fixed with https://github.com/wso2/carbon-identity-framework/pull/6119/commits/931588f41d06bd38d14dd21ecff6b83e13fcc841

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOFacade.java [hunk 11]
```diff
@@ -0,0 +1,313 @@
...
+                ActionPropertyResolverFactory.getActionPropertyResolver(actionDTO.getType());
+
+        if (actionPropertyResolver != null) {
+            properties = actionPropertyResolver.addProperties(actionDTO,
+                    IdentityTenantUtil.getTenantDomain(tenantId));
+        }
+        if (properties != null) {
+            actionDTO.setProperties(properties.entrySet().stream()
+                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue)));
+        }
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOFacade.java:249]: IMO, the objective of the resolver should be to populate the properties.
So either let them to return properties or let them to populate in the transferred DTO. Should choose one path only

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOFacade.java [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+                ActionPropertyResolverFactory.getActionPropertyResolver(
+                        org.wso2.carbon.identity.action.management.model.Action.ActionTypes.valueOf(actionType));
+        if (actionPropertyResolver == null) {
+            return;
+        }
+
+        for (ActionDTO actionDTO : actionDTOS) {
+            actionDTO.setProperties(actionPropertyResolver.getProperties(actionDTO,
+                    IdentityTenantUtil.getTenantDomain(tenantId)));
+        }
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOFacade.java:265]: We should have a way to let the downstream component to populate the ActionDTOs in a bulk manner. If there are any I/O operations this pattern will be heavy

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOImpl.java [hunk 13]
```diff
@@ -212,424 +171,357 @@ public Map<String, Integer> getActionsCountPerType(Integer tenantId) throws Acti
...
+        Map<String, String> actionEndpointProperties = new HashMap<>();
         try {
-            jdbcTemplate.executeUpdate(ActionMgtSQLConstants.Query.CHANGE_ACTION_STATUS,
+            jdbcTemplate.executeQuery(ActionMgtSQLConstants.Query.GET_ACTION_ENDPOINT_INFO_BY_ID,
+                (resultSet, rowNumber) -> {
+                    actionEndpointProperties.put(
+                            resultSet.getString(ActionMgtSQLConstants.Column.ACTION_ENDPOINT_PROPERTY_NAME),
+                            resultSet.getString(ActionMgtSQLConstants.Column.ACTION_ENDPOINT_PROPERTY_VALUE));
+                    return null;
+                },
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOImpl.java:443]: So, these are now no longer just endpoint related properties right ? These are action related properties. Therefore let's rename the constants for readability

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOImpl.java [hunk 14]
```diff
@@ -18,182 +18,141 @@
...
+
+                    actionDTOS.add(actionBuilder.build());
+                }
             }
-            throw ActionManagementUtil.handleServerException(
-                    ActionMgtConstants.ErrorMessages.ERROR_WHILE_ADDING_ACTION, e);
+
+            return actionDTOS;
+        } catch (SQLException e) {
+            throw new ActionMgtServerException("Error while retrieving Actions Basic information by Action Type.", e);
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOImpl.java:98]: This is not basic info right ? This seems to be populating the full object

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOImpl.java [hunk 15]
```diff
@@ -212,424 +171,357 @@ public Map<String, Integer> getActionsCountPerType(Integer tenantId) throws Acti
...
-            return action;
-        } catch (ActionMgtException | ActionMgtRuntimeException e) {
-            /**
-             * Handling {@link ActionMgtRuntimeException}, which is intentionally thrown to represent underlying
-             * exceptions from the {@link #getActionEndpointConfigById(String, Integer)} method.
-             */
-            throw ActionManagementUtil.handleServerException(
-                    ActionMgtConstants.ErrorMessages.ERROR_WHILE_RETRIEVING_ACTION_BY_ID, e);
+            throw ActionManagementExceptionHandler.handleServerException(
+                    ErrorMessage.ERROR_WHILE_RETRIEVING_ACTIONS_COUNT_PER_TYPE, e);
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOImpl.java:175]: Better follow one pattern in handling these exceptions at the DAO layer

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOImpl.java [hunk 16]
```diff
@@ -212,424 +171,357 @@ public Map<String, Integer> getActionsCountPerType(Integer tenantId) throws Acti
...
-                                    property.getValue());
-                            statement.setString(ActionMgtSQLConstants.Column.ACTION_ENDPOINT_PROPERTY_NAME,
-                                    property.getKey());
+                        for (Map.Entry<String, String> property : actionProperties.entrySet()) {
                             statement.setString(ActionMgtSQLConstants.Column.ACTION_ENDPOINT_UUID, actionId);
                             statement.setInt(ActionMgtSQLConstants.Column.TENANT_ID, tenantId);
+                            statement.setString(ActionMgtSQLConstants.Column.ACTION_ENDPOINT_PROPERTY_NAME,
+                                    property.getKey());
+                            statement.setString(ActionMgtSQLConstants.Column.ACTION_ENDPOINT_PROPERTY_VALUE,
+                                    property.getValue());
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOImpl.java:416]: Let's rename the column name constants as this is now related to properties not just endpoint properties

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOFacade.java (hunk 11 above)
Comment (ashanthamara) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOFacade.java]: Fixed with https://github.com/wso2/carbon-identity-framework/pull/6119/commits/edeb975693380b404a45f24842338951d17c1695

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOImpl.java (hunk 14 above)
Comment (ashanthamara) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOImpl.java]: fixed with https://github.com/wso2/carbon-identity-framework/pull/6119/commits/edeb975693380b404a45f24842338951d17c1695

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOImpl.java (hunk 16 above)
Comment (ashanthamara) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOImpl.java]: Fixed with https://github.com/wso2/carbon-identity-framework/pull/6119/commits/dab9dd75643167cc04aa5d35aec2f044d59a58d4

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOImpl.java (hunk 13 above)
Comment (ashanthamara) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOImpl.java]: Fixed with https://github.com/wso2/carbon-identity-framework/pull/6119/commits/dab9dd75643167cc04aa5d35aec2f044d59a58d4

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOFacade.java (hunk 12 above)
Comment (ashanthamara) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOFacade.java]: Fixed with Fixed with https://github.com/wso2/carbon-identity-framework/pull/6119/commits/dab9dd75643167cc04aa5d35aec2f044d59a58d4

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOFacade.java [hunk 17]
```diff
@@ -0,0 +1,414 @@
...
+                // Resolve action properties
+                ActionDTO resolvedUpdatingActionDTO =
+                        getActionDTOWithResolvedUpdatingProperties(updatingActionDTOBuilder.build(), existingActionDTO,
+                                tenantId);
+
+                actionManagementDAO.updateAction(resolvedUpdatingActionDTO, existingActionDTO, tenantId);
+                return null;
+            });
+        } catch (TransactionException e) {
+            handleActionPropertyResolverClientException(e.getCause());
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOFacade.java:141]: This seems to be catching a TransactionException. In that case looks like the logic within this method will not be executed

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/exception/ActionMgtException.java [hunk 18]
```diff
@@ -31,16 +31,9 @@ public ActionMgtException(String message) {
...
     }
 
-    public ActionMgtException(String message, String errorCode) {
-
-        super(message);
-        this.errorCode = errorCode;
-    }
-
-    public ActionMgtException(String message, String errorCode, Throwable cause) {
+    public ActionMgtException(String message, Throwable cause) {
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/exception/ActionMgtException.java:34]: Why was this needed ?

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/service/ActionPropertyResolver.java [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+     * This method is responsible for performing necessary CRUD operations for the properties that need to be added
+     * using other external services.
+     * The Action Management Service ensures that only the properties returned by this method are stored.
+     *
+     * @param actionDTO    ActionDTO object.
+     * @param tenantDomain Tenant domain.
+     * @return ActionDTO object with resolved properties.
+     * @throws ActionPropertyResolverException If an error occurs while resolving the properties.
+     */
+    default ActionDTO resolveAddingProperties(ActionDTO actionDTO, String tenantDomain)
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/service/ActionPropertyResolver.java:46]: Shall we add a better name.
Like populatePropertiesForAddOperation, populateDTOForAddOperation likewise

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/service/impl/ActionManagementServiceImpl.java [hunk 20]
```diff
@@ -38,26 +43,22 @@
...
     private static final ActionValidator ACTION_VALIDATOR = new ActionValidator();
     private static final ActionManagementAuditLogger auditLogger = new ActionManagementAuditLogger();
-    private static final ActionSecretProcessor ACTION_SECRET_PROCESSOR = new ActionSecretProcessor();
 
-    private ActionManagementServiceImpl() {
-    }
+    private final ActionManagementDAOFacade daoFacade;
 
-    public static ActionManagementService getInstance() {
+    public ActionManagementServiceImpl(ActionManagementDAO actionManagementDAO) {
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/service/impl/ActionManagementServiceImpl.java:59]: IMO, we should not pass a DAO for the service layer

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOFacade.java (hunk 17 above)
Comment (ashanthamara) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOFacade.java:141]: The withTransaction() method in jdbctemplate class wraps every exception with a TransactionException. 
If we want to check the exact cause we can check it using the e.getCause().

Wrote unit tests also to validate this scenario.
https://github.com/wso2/carbon-identity-framework/pull/6119/files#diff-b094e19b41079e4bf3e4f26aadb36b1de3e74044332b125c03222d2e8a462fbaR151-R166:~:text=testAddActionWithActionPropertyResolverClientException

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/exception/ActionMgtException.java (hunk 18 above)
Comment (ashanthamara) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/exception/ActionMgtException.java:34]: To facilitate the requirement in ActionMgtServerException(String message, Throwable cause)
https://github.com/wso2/carbon-identity-framework/pull/6119/files#diff-2ca8badd17c1432b432842ffdf47a90fcc5e04d4dba336dc1539d6548af7f218R31:~:text=public%20ActionMgtServerException(String%20message%2C%20Throwable%20cause)%20%7B

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOFacade.java [hunk 21]
```diff
@@ -0,0 +1,418 @@
...
+    public void addAction(ActionDTO actionDTO, Integer tenantId) throws ActionMgtException {
+
+        NamedJdbcTemplate jdbcTemplate = new NamedJdbcTemplate(IdentityDatabaseUtil.getDataSource());
+        try {
+            jdbcTemplate.withTransaction(template -> {
+                ActionDTOBuilder actionDTOBuilder = new ActionDTOBuilder(actionDTO);
+                // Encrypt authentication secrets
+                encryptAddingAuthSecrets(actionDTOBuilder);
+                // Resolve action properties
+                ActionDTO resolvedActionDTO = getActionDTOWithResolvedAddingProperties(actionDTOBuilder.build(),
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOFacade.java:76]: Shall we rename this method for a better name ?
May be -> getResolvedActionDTOWithPropertiesForAddOperation or getResolvedActionDTOForAddOperation

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/exception/ActionDTOModelResolverException.java [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+ * This exception is thrown when there is an issue in performing action type specific operations.
+ */
+public class ActionDTOModelResolverException extends Exception {
+
+    private String description;
+
+    public ActionDTOModelResolverException(String message) {
+
+        super(message);
+    }
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/exception/ActionDTOModelResolverException.java:32]: Do we just accept an exception with a message or do we always expect the message and the description ?

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/exception/ActionDTOModelResolverServerException.java [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+/**
+ * Server Exception class for ActionDTO Model Resolver.
+ * This exception is thrown when there is an issue in performing action type specific operations in the system.
+ */
+public class ActionDTOModelResolverServerException extends ActionDTOModelResolverException {
+
+    public ActionDTOModelResolverServerException(String message) {
+
+        super(message);
+    }
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/exception/ActionDTOModelResolverServerException.java:30]: Same comment above applies here

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/exception/ActionMgtException.java (hunk 18 above)
Comment (malithie) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/exception/ActionMgtException.java:34]: Is that needed and used ?

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/exception/ActionMgtServerException.java [hunk 24]
```diff
@@ -23,29 +23,18 @@
  */
 public class ActionMgtServerException extends ActionMgtException {
 
-    public ActionMgtServerException(String message, String errorCode) {
-
-        super(message, errorCode);
-    }
-
-    public ActionMgtServerException(String message, String description, String errorCode) {
+    public ActionMgtServerException(String message) {
```
Comment (malithie) [inline components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/exception/ActionMgtServerException.java:26]: Check if this function is really needed

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/exception/ActionDTOModelResolverServerException.java (hunk 23 above)
Comment (ashanthamara) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/exception/ActionDTOModelResolverServerException.java]: Fixed with https://github.com/wso2/carbon-identity-framework/pull/6119/commits/79930933227872d78a9a2d01f278b50507184b99

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/exception/ActionMgtServerException.java (hunk 24 above)
Comment (ashanthamara) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/exception/ActionMgtServerException.java:26]: We have one usage for this. Hence keeping it as it is

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/service/impl/ActionManagementServiceImpl.java (hunk 20 above)
Comment (ashanthamara) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/service/impl/ActionManagementServiceImpl.java]: Fixed

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/service/ActionPropertyResolver.java (hunk 19 above)
Comment (ashanthamara) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/service/ActionPropertyResolver.java]: Fixed with https://github.com/wso2/carbon-identity-framework/pull/6119/commits/2d1289a369498b88e58048181ecc730cf3265997

File: components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/exception/ActionMgtException.java (hunk 18 above)
Comment (ashanthamara) [reply components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/exception/ActionMgtException.java:34]: Currently we are utlizing this to build exceptions in the ActionManagementDAO. We can remove this if we introduce new Exception type for the DAO. Will track that separately

**Review by malithie — APPROVED** [review]

**Review by malithie — APPROVED** [review]
